### PR TITLE
multiple sonar instance config

### PR DESCRIPTION
**Pull request by fidelcoria-aa**
**Affects:**all versions.

---

Implement multiple sonar instance support. Note: UI admin page configuration overrides the properties file config.

Refactored client to use method for `setServerDetails` instead of setting in constructor to use injected settings.

### CODE REVIEW

**Comment by rvema** [comment]
@fidelcoria-aa can you resolve the conflicts, also while moving the metrics, can we check if the metrics are mentioned in the properties then they should take precedence over the ones defined in the client classes.

**Comment by fidelcoria-aa** [comment]
I don't think in this case having properties file override metrics is a good idea, because of the reordering bug with mongo. We can discuss further in PR https://github.com/Hygieia/hygieia-codequality-sonar-collector/pull/6

**Comment by Courtneyp123** [comment]
Hi @fidelcoria-aa this PR still makes use of getSonarVersions, instead of getSonarVersion from the new PR.  Could you fix this and the other conflicts?  Thanks in advance.

**Comment by fidelcoria-aa** [comment]
I'll resolve conflicts again once my other PRs are merged.

Backwards compatibility... do you mean changing the properties value from `username` to `usernames`? (`passwords`, too). I think making the name semantically consistent (adding the `s`) with the new type (`List`) is worth breaking backwards compatibility.

BTW changing the type from `String` to `List<String>`  does not break backward compatibility. 
Am I missing anything else?

**Comment by Courtneyp123** [comment]
I can get back to you on backwards compatibility.  For now, your other PRs were merged.  Could you update this branch please?

**Comment by Courtneyp123** [comment]
@fidelcoria-aa I'm getting IndexOutOfBoundsExceptions when I run the new code changes from this PR.  Would we be able to set up a call to look into it?

**Comment by fidelcoria-aa** [comment]
Sure. I'll message you over gitter for details.

**Comment by fidelcoria-aa** [comment]
Ok. so after much refactoring:
Multiple sonar instances can be configured via Admin UI xor properties file, but not both.
Properties file config is ignored if any Admin UI configuration is created.
Passwords or tokens can be used in properties file. Passwords override tokens.
Credentials are optional.

Included logic to avoid IndexOutOfBoundsExceptions (getFromListSafely).

File: src/main/java/com/capitalone/dashboard/collector/DefaultSonar6Client.java [hunk 1]
```diff
@@ -64,34 +66,44 @@
...
         if (!StringUtils.isEmpty(settings.getMetrics63andAbove())) {
             metrics = settings.getMetrics63andAbove();
+        } else {
+            metrics = DEFAULT_METRICS;
+        }
+    }
+
+    @Override
+    public void setServerCredentials(String username, String password, String token) {
+        // use token when given
```
Comment (rvema) [inline src/main/java/com/capitalone/dashboard/collector/DefaultSonar6Client.java]: we need to handle the if both the information is given then provide an error stating only 1 mode of authentication is needed. either token or username/password

File: src/main/java/com/capitalone/dashboard/collector/SonarCollectorTask.java [hunk 2]
```diff
@@ -27,13 +29,7 @@
...
 import org.springframework.stereotype.Component;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (rvema) [inline src/main/java/com/capitalone/dashboard/collector/SonarCollectorTask.java]: can we remove the bulk import and import only specific classes that we need

File: src/main/java/com/capitalone/dashboard/collector/SonarCollectorTask.java [hunk 3]
```diff
@@ -99,12 +116,17 @@ public void collect(SonarCollector collector) {
...
+
+                String username = getFromListSafely(sonarSettings.getUsernames(), i);
+                String password = getFromListSafely(sonarSettings.getPasswords(), i);
+                String token = getFromListSafely(sonarSettings.getTokens(), i);
+                sonarClient.setServerCredentials(username, password, token);
+
+                List<SonarProject> projects = sonarClient.getProjects(instanceUrl);
                 latestProjects.addAll(projects);
 
                 int projSize = ((CollectionUtils.isEmpty(projects)) ? 0 : projects.size());
```
Comment (rvema) [inline src/main/java/com/capitalone/dashboard/collector/SonarCollectorTask.java]: you can use CollectionUtils.size() which is nullsafe and checks for empty as well

**Pull request merged**